boa/src/builtins/array/mod.rs

```diff
@@ -957,6 +957,76 @@ impl Array {
         Ok(Value::from(false))
     }
 
+    /// `Array.prototype.reduce( callbackFn [ , initialValue ] )`
+    ///
+    /// The reduce method traverses left to right starting from the first defined value in the array,
+    /// accumulating a value using a given callback function. It returns the accumulated value.
+    ///  
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array.prototype.reduce
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce
+    pub(crate) fn reduce(
+        this: &Value,
+        args: &[Value],
+        interpreter: &mut Interpreter,
+    ) -> ResultValue {
+        let this = interpreter.to_object(this)?;
+        let callback = match args.get(0) {
+            Some(value) if value.is_function() => value,
+            _ => {
+                return interpreter.throw_type_error(
+                    "missing callback when calling function Array.prototype.reduce",
+                )
+            }
+        };
+        let initial_value = args.get(1).cloned().unwrap_or_else(Value::undefined);
+        let mut length = i32::from(&this.get_field("length"));
+        if length == 0 && initial_value.is_undefined() {
+            return interpreter
+                .throw_type_error("Array contains no elements and initial value is not provided");
+        }
+
+        let mut k = 0;
+        let mut accumulator = if initial_value.is_undefined() {
+            let mut k_present = false;
+            while k < length {
+                if this.has_field(&k.to_string()) {
+                    k_present = true;
+                    break;
+                }
+                k += 1;
+            }
+            if !k_present {
+                return interpreter.throw_type_error(
+                    "Array contains no elements and initial value is not provided",
+                );
+            }
+            let result = this.get_field(k.to_string());
+            k += 1;
+            result
+        } else {
+            initial_value
+        };
+        while k < length {
+            if this.has_field(&k.to_string()) {
+                let arguments = [
+                    accumulator,
+                    this.get_field(k.to_string()),
+                    Value::integer(k),
+                    this.clone(),
+                ];
+                accumulator = interpreter.call(&callback, &Value::undefined(), &arguments)?;
+                // reduce may change the length of the array
+                length = min(length, i32::from(&this.get_field("length")));
```

Comment:
```suggestion
                // reduce may change the length of the array
                length = min(length, interpreter.to_length(&this.get_field("length"))?);
```
also, I understand the updating of length, but I don't see this in the spec. Am I missing something?

Reply:
For some reason resizing the array by setting the length does not cause array accesses beyond the bounds of the array to be undefined. This means `has_field` can still return `true` for an index beyond `arr.length - 1`. IMO, this is probably a bug.

```javascript
let arr = [1, 2, 3, 4, 5];
arr.length = 3;
console.log(arr[4]); // should be undefined, but instead is 5 
```

This means that for now, I'll have to manually update the length. Even if array accesses worked as intended, it might be a good idea to keep track of length to prevent unnecessary iteration.

Reply:
I see. I believe that it a bug. Is because the `internal object methods` are not set right, see https://tc39.es/ecma262/#sec-array-exotic-objects-defineownproperty-p-desc

> Even if array accesses worked as intended, it might be a good idea to keep track of length to prevent unnecessary iteration.

In that case we should leave it as is, a more descriptive comment would be nice to have :)

Reply:
Also here we should use the `.to_lenght()` method:
```
length = min(length, interpreter.to_length(&this.get_field("length"))?);
```